### PR TITLE
Benchmark repo against flashattention-3

### PR DESCRIPTION
**Pull request by MagellaX**
Add key padding mask support and autograd for the Triton fused attention kernel.

The Triton kernel previously lacked a backward pass and key padding mask support, limiting its use to inference. This PR introduces a key padding mask to the Triton kernel and an autograd wrapper that uses the Triton forward and recomputes gradients via PyTorch's SDPA, allowing the fused kernel to be used in training loops while still leveraging its speed for the forward pass.

---
<a href="https://cursor.com/background-agent?bcId=bc-ff90863b-b05e-44d1-a559-52f7a4cdae88">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/open-in-cursor-dark.svg">
    <source media="(prefers-color-scheme: light)" srcset="https://cursor.com/open-in-cursor-light.svg">
    Open in Cursor
  </picture>
</a>
<a href="https://cursor.com/agents?id=bc-ff90863b-b05e-44d1-a559-52f7a4cdae88">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/open-in-web-dark.svg">
    <source media="(prefers-color-scheme: light)" srcset="https://cursor.com/open-in-web-light.svg">
    Open in Web
  </picture>
</a>

    

---

## Summary by cubic
Add key padding mask support and an autograd path to the Triton fused online attention. This enables training with masked inputs while keeping the fast Triton forward.

- **New Features**
  - Triton kernel now supports a key padding mask [B, S_k] alongside causal masking; LSE output is preserved.
  - Added an autograd wrapper: forward runs on Triton, gradients are recomputed via PyTorch SDPA for training.
  - Forward routing prefers Triton on CUDA when mask is [B, S_k] (or None) and dropout is 0; otherwise falls back to SDPA.
  - SDPA path safely combines padding mask and causal by using an additive mask; API remains unchanged.

### CODE REVIEW

**Comment by MagellaX** [comment]
Alright now, re-run the review @cubic-dev-ai

**Comment by MagellaX** [comment]
@cubic-dev-ai re-run the review

**Comment by MagellaX** [comment]
@cursoragent make the required changes according to the cubic-dev-ai bot

**Comment by MagellaX** [comment]
Re-run the review @cubic-dev-ai
